api/v1alpha1/stage_status_types.go

```diff
@@ -0,0 +1,87 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"time"
+)
+
+// StageStatusPhase defines the observed stage phase
+// +kubebuilder:validation:Enum={Progressing,Succeeded,Failed}
+type StageStatusPhase string
+
+const (
+	PhaseProgressing StageStatusPhase = "Progressing"
+	PhaseSucceeded   StageStatusPhase = "Succeeded"
+	PhaseFailed      StageStatusPhase = "Failed"
+)
+
+// StageStatus defines the observed stage status
+type StageStatus struct {
+	Name       string           `json:"name"`
+	Phase      StageStatusPhase `json:"stage,omitempty"`
+	Message    string           `json:"message,omitempty"`
+	Targets    []string         `json:"targets,omitempty"`
+	Syncing    []string         `json:"syncing,omitempty"`
+	Requeued   []string         `json:"requeued,omitempty"`
+	Failed     []string         `json:"failed,omitempty"`
+	Completed  []string         `json:"completed,omitempty"`
+	StartedAt  metav1.Time      `json:"startedAt,omitempty"`
+	FinishedAt metav1.Time      `json:"finishedAt,omitempty"`
+}
+
+// GetStageStatus returns a pointer to the Status.Stages slice
+func (in *ProgressiveRollout) GetStageStatus() *[]StageStatus {
+	return &in.Status.Stages
+}
+
+// NewStageStatus adds a new StageStatus
+func (in *ProgressiveRollout) NewStageStatus(n, m string, p StageStatusPhase, t, s, r, f, c []string) StageStatus {
```

Comment:
why is completed for instance a list of strings? I am still getting used to golang conventions, but it is quite hard to navigate all these single letter variables, could we make them more descriptive?

Reply:
Because of the output we discussed on https://github.com/Skyscanner/argocd-progressive-rollout/issues/8? If not a list of string, what would you use?

Reply:
You are correct, we should use more descriptive parameters. I'm going to fix it here and fix as well in https://github.com/Skyscanner/argocd-progressive-rollout/blob/main/api/v1alpha1/progressiverollout_types.go#L83 that is the only other function that uses single letter names.